File: Models/Reactor.cs
using BioDieselProject.Interfaces;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BioDieselProject.Entity
{
    public class Reactor : IMachines
    {
        public double naOh = 0;
        public double etOh = 0;
        public double oil = 0;

        public Reactor()
        {
            Capacity = 0;
            Flow = 5;
            Volume = 5;
        }

        // metodo adiciona a quantidade que for possivel da substancia, se possivel adicionar tudo ele retorna o valor do que sobrou
        //da substancia
        public object setCapacityNaoh(double quantity)
        {
            if (Capacity <= Volume)
            {
                double sobrou = Volume - Capacity;
                if (quantity <= sobrou)
                {
                    Capacity += quantity;
                    naOh += quantity;
                    quantity = 0;
                }
                else
                {
                    Capacity += sobrou;
                    naOh += sobrou;
                    quantity -= sobrou;
                }
            }
            return new
            {
                quantity
            };
        }

        public object setCapacityEtoh(double quantity)
        {
            if (Capacity <= Volume)
            {
                double sobrou = Volume - Capacity;
                if (quantity <= sobrou)
                {
                    Capacity += quantity;
                    etOh += quantity;
                    quantity = 0;
                }
                else
                {
                    Capacity += sobrou;
                    etOh += sobrou;
                    quantity -= sobrou;
                }
            }
            return new
            {
                quantity
            };
        }

        public object setCapacityOil(double quantity)
        {
            if (Capacity <= Volume)
            {
                double sobrou = Volume - Capacity;
                if (quantity <= sobrou)
                {
                    Capacity += quantity;
                    oil += quantity;
                    quantity = 0;
                }
                else
                {
                    Capacity += sobrou;
                    oil += sobrou;
                    quantity -= sobrou;
                }
            }
            return new
            {
                quantity
            };
        }

        // Verificargit clone
        public override object trasfer()
        {
            double transfer = 0;
            if (Capacity > 0)
            {
                if (Capacity <= Flow)
                {
                    double parte = Capacity/4;
                    if (naOh >= parte && etOh >= parte && oil >= (parte * 2))
                    {
                        transfer = Capacity;
                        Capacity -= transfer;
                        naOh -= parte;
                        etOh -= parte;
                        oil -= (parte*2);
                    }
                }
                else
                {
                    double parte = Flow / 4;
                    if (naOh >= parte && etOh >= parte && oil >= (parte * 2))
                    {
                        transfer = Flow;
                        Capacity -= transfer;
                        naOh -= parte;
                        etOh -= parte;
                        oil -= (parte*2);
                    }
                }
            }
            return new { transfer, naOh, etOh, oil };
        }
    }
}

File: Models/Reactor.py
import threading

from Interface.IMachines import IMachines
import socket
import time
import select

class Reactor(IMachines):

    def __init__(self):
        super().__init__()
        self.Capacity = 0
        self.Flow = 5
        self.Volume = 5
        self.naOh = 0
        self.etOh = 0
        self.oil = 0
        self.host = ""
        self.portToDecantador = 65440
        self.port = 65441

    def getRestante(self) -> int:
        return self.Volume - self.Capacity

    def setCapacityNaoh(self, quantity):
        self.Capacity += quantity
        self.naOh += quantity

    def setCapacityEtoh(self,  quantity):
        self.Capacity += quantity
        self.etOh += quantity

    def setCapacityOil(self, quantity):
        self.Capacity += quantity
        self.oil += quantity

    def calculateTransfer(self, restante):
        transfer = 0
        parte = 0
        if (self.Capacity > 0):
            sizeSubstance = self.Capacity if self.Capacity <= self.Flow else self.Flow
            quantidadeASerTransferida = restante if restante <= sizeSubstance else sizeSubstance
            parte = quantidadeASerTransferida/4
            if (self.naOh >= parte and self.etOh >= parte and self.oil >= (parte * 2)):
                transfer = quantidadeASerTransferida
                self.Capacity -= transfer
                self.naOh -= parte
                self.etOh -= parte
                self.oil -= (parte*2)

        return transfer

    def transfereLoop(self):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect((self.host, self.portToDecantador))
            while True:
                if self.Capacity > 0:
                    s.send(b"get_restante")
                    data = s.recv(1024).decode("utf-8")
                    transfer = self.calculateTransfer(float(data))
                    if transfer > 0:
                        sendString = f"set_capacity {transfer}"
                        s.send(sendString.encode("utf-8"))
                    time.sleep(1)

    def responde(self, clientsocket, addr):
        timeout = 30
        while True:
            ready_sockets, _, _ = select.select(
                [clientsocket], [], [], timeout
            )
            if ready_sockets:
                receivedMessage = clientsocket.recv(1024).decode("utf-8")
                receivedMessage = receivedMessage.split()
                if receivedMessage[0] == "get_restante":
                    restante = str(self.getRestante())
                    print(f"\nRetornando valor: {restante}\n")
                    clientsocket.send(restante.encode("utf-8"))
                elif receivedMessage[0] == "set_oil":
                    self.setCapacityOil(float(receivedMessage[1]))
                elif receivedMessage[0] == "set_etoh":
                    self.setCapacityEtoh(float(receivedMessage[1]))
                elif receivedMessage[0] == "set_naoh":
                    self.setCapacityNaoh(float(receivedMessage[1]))
                elif receivedMessage[0] == "get_oil":
                    sendCapacidadeOil = f"{self.oil}"
                    clientsocket.send(sendCapacidadeOil.encode("utf-8"))
                elif receivedMessage[0] == "get_etoh":
                    sendCapacidadeEtoh = f"{self.etOh}"
                    clientsocket.send(sendCapacidadeEtoh.encode("utf-8"))
                elif receivedMessage[0] == "get_naoh":
                    sendCapacidadeNaoh = f"{self.naOh}"
                    clientsocket.send(sendCapacidadeNaoh.encode("utf-8"))

        clientsocket.close()

    def verify(self):

        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.bind((self.host, self.port))
            s.listen(5)
            while True:
                conn, addr = s.accept()
                threading.Thread(target=self.responde, args=(conn, addr)).start()
